lgc/patch/PatchFetchShader.cpp

```diff
@@ -0,0 +1,294 @@
+
+  auto &entryArgIdxs = intfData->entryArgIdxs;
+  entryArgIdxs.initialized = true;
+  entryArgIdxs.vs.baseVertex = vsInterfaceData->getBaseVertexRegister();
+  intfData->userDataUsage.vs.baseVertex = vsInterfaceData->getBaseInstanceRegister();
+  entryArgIdxs.vs.baseInstance = vsInterfaceData->getBaseInstanceRegister();
+  intfData->userDataUsage.vs.baseInstance = vsInterfaceData->getBaseInstanceRegister();
+
+  // Add system values.  The return types are changed to Float32 to indicate that they must be in a VGPR on exit.
+
+  // Vertex ID
+  builtInUsage.vertexIndex = true;
+  entryArgIdxs.vs.vertexId = argTys.size();
+  argTys.push_back(Type::getInt32Ty(*m_context));
+  retTys.push_back(Type::getFloatTy(*m_context));
+
+  // Relative vertex ID (auto index)
+  entryArgIdxs.vs.relVertexId = argTys.size();
+  argTys.push_back(Type::getInt32Ty(*m_context));
+  retTys.push_back(Type::getFloatTy(*m_context));
+
+  // Primitive ID
+  builtInUsage.primitiveId = true;
+  entryArgIdxs.vs.instanceId = argTys.size();
+  entryArgIdxs.vs.primitiveId = argTys.size();
+  argTys.push_back(Type::getInt32Ty(*m_context));
+  retTys.push_back(Type::getFloatTy(*m_context));
+
+  // Instance ID
+  builtInUsage.instanceIndex = true;
+  entryArgIdxs.vs.instanceId = argTys.size();
+  argTys.push_back(Type::getInt32Ty(*m_context));
+  retTys.push_back(Type::getFloatTy(*m_context));
+
+  // Add the vertex input types to the return struct.
+  const auto& vsInputTypeInfo = vsInterfaceData->getVertexInputTypeInfo();
+  for (auto vertexInputInfo : vsInputTypeInfo) {
+    unsigned location = vertexInputInfo.first;
+    Type *type = vsInterfaceData->getVertexInputType(location, m_context);
+
+    // To make sure that these values are returned in VGPRs, we must cast the type to a floating point type.
+    if (!type->getScalarType()->isFloatTy()) {
+      if (type->isVectorTy()) {
+        type = VectorType::get(Type::getFloatTy(*m_context), type->getVectorElementCount());
+      } else {
+        type = Type::getFloatTy(*m_context);
+      }
+    }
+    retTys.push_back(type);
+  }
+
+  Type *returnStructType = StructType::create(*m_context, retTys);
+  return FunctionType::get(returnStructType, argTys, false);
+}
+
+// =====================================================================================================================
+// Load the vertex input at the given location and return that value.
+//
+// @param fetchType [in]: The type of the vertex input at the given location.
+// @param returnType [in]: The type that the returned value should be.
+// @param location [in]: The location of the vertex input to load.
+// @param insertPos [in]: Newly generated code will be inserted before this instruction.
+// @param vertexFetch [in]: An already initialzed VertexFetch object.
+Value *PatchFetchShader::fetchVertexInput(Type *fetchType, Type *returnType, unsigned location, Instruction *insertPos,
+                                          VertexFetch *vertexFetch) {
```

Comment:
Please change this method to take a builder argument and use that for the bitcast.

Changing vertexFetch->run to take a builder can happen later

Reply:
I have the builder available as a member variable, so I use that to get the insertion point, and to build the bit cast.  Is that good enough?  I do not pass it as a parameter.

Reply:
Yes, that sounds fine.